File: packages/example_app/lib/features/order/domain/usecases/order_usecases.dart
import 'package:dartz/dartz.dart';

import '../../../../core/errors/failures.dart';
import '../../../../core/usecase/usecase.dart';
import '../entities/order_entity.dart';
import '../repositories/order_repository.dart';

class GetAllOrders extends UseCase<List<OrderEntity>, NoParams> {
  const GetAllOrders(this._repo);
  final OrderRepository _repo;

  @override
  Future<Either<Failure, List<OrderEntity>>> call(NoParams params) =>
      _repo.getAll();
}

class GetOrderById extends UseCase<OrderEntity, String> {
  const GetOrderById(this._repo);
  final OrderRepository _repo;

  @override
  Future<Either<Failure, OrderEntity>> call(String id) => _repo.getById(id);
}

class AddOrder extends UseCase<OrderEntity, OrderEntity> {
  const AddOrder(this._repo);
  final OrderRepository _repo;

  @override
  Future<Either<Failure, OrderEntity>> call(OrderEntity params) =>
      _repo.add(params);
}

class UpdateOrder extends UseCase<OrderEntity, OrderEntity> {
  const UpdateOrder(this._repo);
  final OrderRepository _repo;

  @override
  Future<Either<Failure, OrderEntity>> call(OrderEntity params) =>
      _repo.update(params);
}

class DeleteOrder extends UseCase<Unit, String> {
  const DeleteOrder(this._repo);
  final OrderRepository _repo;

  @override
  Future<Either<Failure, Unit>> call(String id) => _repo.delete(id);
}

class SearchOrders extends UseCase<List<OrderEntity>, String> {
  const SearchOrders(this._repo);
  final OrderRepository _repo;

  @override
  Future<Either<Failure, List<OrderEntity>>> call(String query) =>
      _repo.search(query);
}

File: packages/example_react_app/src/features/order/domain/usecases/order.usecases.ts
/// Use cases for Order.
import type { Either } from '../../../../core/usecase/usecase';
import type { Failure } from '../../../../core/errors/failures';
import type { OrderEntity } from '../entities/order.entity';
import type { OrderRepository } from '../repositories/order.repository';
import type { PaginationParams, PaginatedResponse } from '../../../../core/pagination/pagination';

export class GetAllOrdersUseCase {
  private repository: OrderRepository;

  constructor(repository: OrderRepository) {
    this.repository = repository;
  }

  async execute(): Promise<Either<Failure, OrderEntity[]>> {
    return this.repository.getAll();
  }
}

export class GetOrderByIdUseCase {
  private repository: OrderRepository;

  constructor(repository: OrderRepository) {
    this.repository = repository;
  }

  async execute(id: string): Promise<Either<Failure, OrderEntity>> {
    return this.repository.getById(id);
  }
}

export class AddOrderUseCase {
  private repository: OrderRepository;

  constructor(repository: OrderRepository) {
    this.repository = repository;
  }

  async execute(entity: OrderEntity): Promise<Either<Failure, OrderEntity>> {
    return this.repository.add(entity);
  }
}

export class UpdateOrderUseCase {
  private repository: OrderRepository;

  constructor(repository: OrderRepository) {
    this.repository = repository;
  }

  async execute(entity: OrderEntity): Promise<Either<Failure, OrderEntity>> {
    return this.repository.update(entity);
  }
}

export class DeleteOrderUseCase {
  private repository: OrderRepository;

  constructor(repository: OrderRepository) {
    this.repository = repository;
  }

  async execute(id: string): Promise<Either<Failure, void>> {
    return this.repository.delete(id);
  }
}

export class SearchOrdersUseCase {
  private repository: OrderRepository;

  constructor(repository: OrderRepository) {
    this.repository = repository;
  }

  async execute(query: string): Promise<Either<Failure, OrderEntity[]>> {
    return this.repository.search(query);
  }
}

export class GetAllOrdersPaginatedUseCase {
  private repository: OrderRepository;

  constructor(repository: OrderRepository) {
    this.repository = repository;
  }

  async execute(params: PaginationParams): Promise<Either<Failure, PaginatedResponse<OrderEntity>>> {
    return this.repository.getAllPaginated(params);
  }
}
